x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -1126,20 +1130,19 @@ static void writeAutoBucket(PlanStreamOutput out, AutoBucket bucket) throws IOEx
         out.writeExpression(bucket.to());
     }
 
-    static Case readCase(PlanStreamInput in) throws IOException {
-        return new Case(Source.EMPTY, in.readList(readerFromPlanReader(PlanStreamInput::readExpression)));
-    }
-
-    static void writeCase(PlanStreamOutput out, Case caseValue) throws IOException {
-        out.writeCollection(caseValue.children(), writerFromPlanWriter(PlanStreamOutput::writeExpression));
-    }
+    static final Map<String, BiFunction<Source, List<Expression>, ScalarFunction>> VARARG_CTORS = Map.ofEntries(
+        entry(name(Case.class), Case::new),
+        entry(name(Coalesce.class), Coalesce::new),
+        entry(name(Greatest.class), Greatest::new),
+        entry(name(Least.class), Least::new)
+    );
```

Comment:
question (I'm not fully familiar with the code base yet, so this is going to sound naive): It seems like instead of a map, one could also define an enum representing the vararg functions, or maybe have them all implement an interface that is used by vararg functions. Why do we go with a map?

Reply:
They could all implement an interface, namely `Writeable`. Then we wouldn't have any of these. But we ended up going this way to prevent us from having to implement that interface on QL code. To be honest, I wonder if, now that we've merged to `main`, we should back that out.

Reply:
Implementing Writable means touching all the QL code (expression, attribute, etc..) instead of having one class that handles all the messy details of writing and another for reading.